### ISSUE TITLE
Generated keys may be unsuitable for encryption

### ISSUE THREAD

**Issue by dalai4git** [ProtonMail/go-crypto#184]
In `openpgp/key_generation.go` the `rsa.GenerateKey` function from the standard library is used. As far as I can tell, the new key may have primes such that p > q, which is not according to RFC 4880 and can lead to occasional problems when decrypting using gpg.

See also: https://dev.gnupg.org/T4579

**Comment by jsoo1** [comment]
If I understand this thread/patch correctly (https://dev.gnupg.org/T8171) then I think libgcrypt is going to start rejecting secret keys with p > q.

**Comment by twiss** [comment]
It seems you're correct. And, sorry for not responding to this issue before.
Would either of you be up for making a PR?
(We don't generate RSA keys anymore at Proton.)

**Comment by jsoo1** [comment]
I'm not opposed to it but I am no expert.

**Comment by twiss** [comment]
If you have time please do, we can review it ofc.

**Comment by jam-awake** [comment]
I took a look at this. Would a fix look something like:
1. generate a key like normal
2. check the P and Q primes to ensure the constraint is satisfied
3. if not satisfied, swap the primes and recompute the precomputed values

I think that can be done for normal RSA keys. I'm not sure how this affects multi-prime RSA keys.

**Comment by twiss** [comment]
Yeah, that sounds fine. OpenPGP doesn't support multi-prime RSA keys.

**Comment by jam-awake** [comment]
`gnupg@2.5.19` was released (https://lists.gnupg.org/pipermail/gnupg-announce/2026q2/000504.html):
> Other changes:
>   - agent: Raise an error when p >= q for RSA keys to detect
>     incorrect generated *PGP keys.  [T8171]

We'll see if this issue sees more traffic.
